### PR TITLE
Only consider genpop appeals when attempting to evenly distribute cases to judges

### PR DESCRIPTION
**Pull request by hschallhorn**
Resolves discrepancy found here: https://dsva.slack.com/archives/CN3FQR4A1/p1601297485008600

### Description
Our priority push job runs in 2 phases.
First, we distribute all cases that are tied to a judge through previous decision or a held hearing.
Second, we pull a count of all remaining priority cases ready to be distributed and calculate how to distribute them evenly across judges.
However, when we pull the count of remaining cases, we were previously including cases that are tied to judges but could not be distributed. This could happen for a number of reasons (judge case is tied to no longer works for the board, is not a user in caseflow, is an AVLJ, etc). The issue is that when we try evenly distribute remaining cases, we cannot distribute cases tied to judges to another judge.
This let to us only having 60 cases to distribute, but our calculations were done assuming there would be 350+ cases to distribute.

This PR ensures the calculations done on `ready_priority_appeals_count` do not consider cases we cannot distribute (cases tied to another judge)

### Acceptance Criteria
- [ ] When calculating priority target for the acd priority push job, only consider genpop cases

### Testing Plan
1. in prod
```ruby
PushPriorityAppealsToJudgesJob.new.ready_priority_appeals_count
=> 346
# Or some other 300+ number

# what should this be instead?
query = <<-SQL
  #{VACOLS::CaseDocket::SELECT_PRIORITY_APPEALS}
  where VLJ is null
SQL

VACOLS::CaseDocket.connection.exec_query(query).to_hash.count
=> 34
HearingRequestDistributionQuery.new(base_relation: HearingRequestDocket.new.appeals(ready: true, priority: true), genpop: "only_genpop").call.count
=> 0
DirectReviewDocket.new.count(ready: true, priority: true)
=> 0
EvidenceSubmissionDocket.new.count(ready: true, priority: true)
=> 0
# The sum of these numbers should be MUCH lower than the one returned from ready_priority_appeals_count
```

### CODE REVIEW

File: app/jobs/push_priority_appeals_to_judges_job.rb [hunk 1]
```diff
@@ -123,8 +123,8 @@ def docket_coordinator
     @docket_coordinator ||= DocketCoordinator.new
   end
 
-  def ready_priority_appeals_count
-    @ready_priority_appeals_count ||= docket_coordinator.priority_count
+  def ready_genpop_priority_appeals_count
+    @ready_genpop_priority_appeals_count ||= docket_coordinator.genpop_priority_count
```
Comment (hschallhorn) [inline app/jobs/push_priority_appeals_to_judges_job.rb:127]: Rename!

File: app/models/concerns/proportion_hash.rb [hunk 2]
```diff
@@ -52,7 +52,7 @@ def add_fixed_proportions!(fixed)
 
   def stochastic_allocation(num)
     result = transform_values { |proportion| (num * proportion).floor }
-    rem = num - result.values.reduce(0, :+)
+    rem = num - result.values.sum
```
Comment (hschallhorn) [inline app/models/concerns/proportion_hash.rb:55]: small cleanup

File: app/models/docket.rb [hunk 3]
```diff
@@ -30,6 +30,10 @@ def count(priority: nil, ready: nil)
     appeals(priority: priority, ready: ready).ids.size
   end
 
+  def genpop_priority_count
+    count(priority: true, ready: true)
+  end
+
```
Comment (hschallhorn) [inline app/models/docket.rb:36]: for evidence and direct review docket, all appeals are genpop, so we don't need to specify anything more than "ready" and "priority". This is overridden in [hearing request docket](https://github.com/department-of-veterans-affairs/caseflow/pull/15328/files#diff-2079305d0bbcd85adf95a8923969ef6fR18)

File: app/models/docket_coordinator.rb [hunk 4]
```diff
@@ -100,7 +100,11 @@ def priority_count
     @priority_count ||= dockets
       .values
       .map { |docket| docket.count(priority: true, ready: true) }
-      .reduce(0, :+)
+      .sum
+  end
+
+  def genpop_priority_count
+    @genpop_priority_count ||= dockets.values.map(&:genpop_priority_count).sum
```
Comment (hschallhorn) [inline app/models/docket_coordinator.rb:107]: New function to pull all genpop_priority_count from each docket and sum them up!

File: app/models/dockets/hearing_request_docket.rb [hunk 5]
```diff
@@ -5,14 +5,20 @@ def docket_type
...
 
+  def age_of_n_oldest_priority_appeals(num)
     HearingRequestDistributionQuery.new(
-      base_relation: relation, genpop: "only_genpop"
+      base_relation: ready_priority_appeals.limit(num), genpop: "only_genpop"
     ).call.map(&:ready_for_distribution_at)
   end
 
+  def genpop_priority_count
+    HearingRequestDistributionQuery.new(base_relation: ready_priority_appeals, genpop: "only_genpop").call.count
```
Comment (hschallhorn) [inline app/models/dockets/hearing_request_docket.rb:19]: grabs all ready priority appeals that are [not tied to a judge ](https://github.com/department-of-veterans-affairs/caseflow/blob/55b1e7c184adb9147195a5158f526964c5436c29/app/queries/hearing_request_distribution_query.rb#L36-L41)

File: app/models/dockets/legacy_docket.rb [hunk 6]
```diff
@@ -22,12 +22,16 @@ def count(priority: nil, ready: nil)
...
+    LegacyAppeal.repository.genpop_priority_count
+  end
+
   def weight
     count(priority: false) + nod_count * NOD_ADJUSTMENT
   end
 
   def ready_priority_appeal_ids
-    VACOLS::CaseDocket.priority_ready_appeal_vacols_ids
+    LegacyAppeal.repository.priority_ready_appeal_vacols_ids
```
Comment (hschallhorn) [inline app/models/dockets/legacy_docket.rb:34]: Unrelated change. Pulled out to [appeal repository](https://github.com/department-of-veterans-affairs/caseflow/pull/15328/files#diff-fd86a0d6982fc8d827021423d942eab4R716) so we can gather metrics

File: spec/models/docket_coordinator_spec.rb [hunk 7]
```diff
@@ -1,210 +1,283 @@
...
-    let(:nonpriority_decisions_per_year) { 1340 }
-    let(:nonpriority_legacy_count) { 80 }
-
-    it "uses the pacesetting direct review proportion" do
-      expect(docket_coordinator.docket_proportions).to include(
-        legacy: 0.8,
-        evidence_submission: 0.05,
-        hearing: 0.05
-      )
-      expect(docket_coordinator.interpolated_minimum_direct_review_proportion).to be_within(0.001).of(0.1)
```
Comment (hschallhorn) [inline spec/models/docket_coordinator_spec.rb:207]: Everything above this was pulled into its own context so we don't run all that uneeded set up again for our tests

File: spec/models/docket_coordinator_spec.rb [hunk 8]
```diff
@@ -1,210 +1,283 @@
 # frozen_string_literal: true
 
 describe DocketCoordinator do
-  before do
```
Comment (hschallhorn) [inline spec/models/docket_coordinator_spec.rb:4]: Whitespace!

File: spec/models/docket_coordinator_spec.rb (hunk 7 above)
Comment (hschallhorn) [reply spec/models/docket_coordinator_spec.rb:207]: Actually new tests are below

File: app/models/vacols/case_docket.rb [hunk 9]
```diff
@@ -166,6 +166,15 @@ def self.counts_by_priority_and_readiness
   end
   # rubocop:enable Metrics/MethodLength
 
+  def self.genpop_priority_count
+    query = <<-SQL
+      #{SELECT_PRIORITY_APPEALS}
+      where VLJ is null
+    SQL
```
Comment (hschallhorn) [inline app/models/vacols/case_docket.rb:173]: grabs all ready priority appeals that do not have a [linked VLJ though prior decisions or though a hearing](https://github.com/department-of-veterans-affairs/caseflow/blob/ca8c6b78483fa56d2fd300b64b51aebb0e423bbe/app/models/vacols/case_docket.rb#L111)

**Review by lomky — APPROVED** [review]
Couple nice-to-have comments, nothing blocking! :beetle: :hammer: :tada:

**Comment by lomky** [comment]
I disagree with the codeclimate alert here - we're overriding behavior, where else should it go? :thinking:
